File: tracking/tracker.cpp
#include <opencv2/dnn.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/video/tracking.hpp>
#include <opencv2/videoio.hpp>

// KCF and CSRT live in opencv_contrib when that optional header is installed.
#if __has_include(<opencv2/tracking.hpp>)
#include <opencv2/tracking.hpp>
#define LEARNOPENCV_HAS_CONTRIB_TRACKING 1
#else
#define LEARNOPENCV_HAS_CONTRIB_TRACKING 0
#endif

#include <algorithm>
#include <cctype>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

// The filesystem alias keeps model-path validation concise and type-safe.
namespace fs = std::filesystem;

// Tracker names are case-insensitive at the command line.
std::string upper(std::string value) {
    std::transform(value.begin(), value.end(), value.begin(),
                   [](unsigned char character) {
                       return static_cast<char>(std::toupper(character));
                   });
    return value;
}

// The downloader accepts lowercase tracker group names.
std::string lower(std::string value) {
    std::transform(value.begin(), value.end(), value.begin(),
                   [](unsigned char character) {
                       return static_cast<char>(std::tolower(character));
                   });
    return value;
}

// Fail before calling OpenCV when a tracker model set is incomplete.
void requireModels(
    const std::string& trackerName,
    const fs::path& modelsDir,
    const std::vector<std::string>& filenames) {
    std::vector<std::string> missing;
    // Report every missing file together so the reader can fix one download.
    for (const std::string& filename : filenames) {
        if (!fs::is_regular_file(modelsDir / filename)) {
            missing.push_back(filename);
        }
    }
    if (missing.empty()) {
        return;
    }

    // Include the exact recovery command in the reader-facing error.
    std::ostringstream message;
    message << trackerName << " requires model files in " << modelsDir
            << "; missing: ";
    for (std::size_t index = 0; index < missing.size(); ++index) {
        if (index != 0) {
            message << ", ";
        }
        message << missing[index];
    }
    message << ". Run: python download_models.py --tracker "
            << lower(trackerName);
    throw std::runtime_error(message.str());
}

// Construct only trackers exposed by the linked OpenCV installation.
cv::Ptr<cv::Tracker> createTracker(
    const std::string& requestedName, const fs::path& modelsDir) {
    const std::string name = upper(requestedName);
#if LEARNOPENCV_HAS_CONTRIB_TRACKING
    // These factories compile only when the contrib tracking header exists.
    if (name == "CSRT") {
        return cv::TrackerCSRT::create();
    }
    if (name == "KCF") {
        return cv::TrackerKCF::create();
    }
#endif
    if (name == "MIL") {
        return cv::TrackerMIL::create();
    }
    if (name == "DASIAMRPN") {
        // DaSiamRPN separates its search network from two template kernels.
        requireModels(
            name, modelsDir,
            {"dasiamrpn_model.onnx", "dasiamrpn_kernel_cls1.onnx",
             "dasiamrpn_kernel_r1.onnx"});
        cv::TrackerDaSiamRPN::Params parameters;
        // Pass absolute or working-directory-relative paths explicitly.
        parameters.model = (modelsDir / "dasiamrpn_model.onnx").string();
        parameters.kernel_cls1 =
            (modelsDir / "dasiamrpn_kernel_cls1.onnx").string();
        parameters.kernel_r1 =
            (modelsDir / "dasiamrpn_kernel_r1.onnx").string();
        // The documented CPU backend is reproducible across OpenCV 4 and 5.
        parameters.backend = cv::dnn::DNN_BACKEND_OPENCV;
        parameters.target = cv::dnn::DNN_TARGET_CPU;
        return cv::TrackerDaSiamRPN::create(parameters);
    }
    if (name == "NANO") {
        // NanoTrack v2 uses one template backbone and one prediction head.
        requireModels(
            name, modelsDir,
            {"nanotrack_backbone_sim.onnx", "nanotrack_head_sim.onnx"});
        cv::TrackerNano::Params parameters;
        parameters.backbone =
            (modelsDir / "nanotrack_backbone_sim.onnx").string();
        parameters.neckhead =
            (modelsDir / "nanotrack_head_sim.onnx").string();
        parameters.backend = cv::dnn::DNN_BACKEND_OPENCV;
        parameters.target = cv::dnn::DNN_TARGET_CPU;
        return cv::TrackerNano::create(parameters);
    }
    if (name == "VIT") {
        // VitTrack packages its compact transformer into one ONNX graph.
        requireModels(
            name, modelsDir, {"object_tracking_vittrack_2023sep.onnx"});
        cv::TrackerVit::Params parameters;
        parameters.net =
            (modelsDir / "object_tracking_vittrack_2023sep.onnx").string();
        parameters.backend = cv::dnn::DNN_BACKEND_OPENCV;
        parameters.target = cv::dnn::DNN_TARGET_CPU;
        return cv::TrackerVit::create(parameters);
    }
    throw std::invalid_argument(
        "Tracker '" + requestedName +
        "' is unavailable. Choose MIL, CSRT, KCF, DASIAMRPN, NANO, or VIT. "
        "CSRT and KCF are capability-gated contrib options.");
}

cv::Rect parseBoundingBox(const std::string& text) {
    std::stringstream stream(text);
    cv::Rect box;
    char separator1 = '\0';
    char separator2 = '\0';
    char separator3 = '\0';
    if (!(stream >> box.x >> separator1 >> box.y >> separator2
          >> box.width >> separator3 >> box.height) ||
        separator1 != ',' || separator2 != ',' || separator3 != ',' ||
        box.width <= 0 || box.height <= 0) {
        throw std::invalid_argument(
            "Bounding box must be x,y,width,height with positive dimensions.");
    }
    stream >> std::ws;
    if (!stream.eof()) {
        throw std::invalid_argument(
            "Bounding box must contain exactly four comma-separated integers.");
    }
    return box;
}

void validateBoundingBox(const cv::Rect& box, const cv::Size& frameSize) {
    if (box.x < 0 || box.y < 0 || box.width <= 0 || box.height <= 0 ||
        box.width > frameSize.width - box.x ||
        box.height > frameSize.height - box.y) {
        throw std::invalid_argument("Bounding box lies outside the first frame.");
    }
}

cv::VideoWriter openWriter(
    const std::string& path, double fps, const cv::Size& frameSize) {
    const bool isMp4 =
        path.size() >= 4 && path.substr(path.size() - 4) == ".mp4";
    const int codec = isMp4
        ? cv::VideoWriter::fourcc('m', 'p', '4', 'v')
        : cv::VideoWriter::fourcc('M', 'J', 'P', 'G');
    cv::VideoWriter writer(path, codec, fps, frameSize);
    if (!writer.isOpened()) {
        throw std::runtime_error("Could not open output video: " + path);
    }
    return writer;
}

}  // namespace

int main(int argc, char** argv) {
    // CommandLineParser supplies defaults that match the bundled Chaplin clip.
    const std::string keys =
        "{help h usage ?||Show this help message}"
        "{input|videos/chaplin.mp4|Input video path}"
        "{tracker|MIL|MIL, CSRT, KCF, DASIAMRPN, NANO, or VIT}"
        "{models-dir|models|Directory containing verified ONNX tracker models}"
        "{bbox|287,23,86,320|Initial x,y,width,height}"
        "{select-roi||Select the initial box interactively}"
        "{output||Optional annotated output video}"
        "{snapshot||Optional final annotated frame}"
        "{max-frames|-1|Maximum update frames; -1 processes all}"
        "{display||Show an interactive window}";
    cv::CommandLineParser parser(argc, argv, keys);
    if (parser.has("help")) {
        parser.printMessage();
        return 0;
    }
    if (!parser.check()) {
        parser.printErrors();
        return 2;
    }

    try {
        const std::string inputPath = parser.get<std::string>("input");
        const std::string trackerName = upper(parser.get<std::string>("tracker"));
        // Keep models outside the executable so the downloader can verify them.
        const fs::path modelsDir =
            fs::path(parser.get<std::string>("models-dir"));
        const std::string outputPath = parser.get<std::string>("output");
        const std::string snapshotPath = parser.get<std::string>("snapshot");
        const int maxFrames = parser.get<int>("max-frames");
        const bool display = parser.has("display");
        const bool selectRoi = parser.has("select-roi");
        if (maxFrames == 0 || maxFrames < -1) {
            throw std::invalid_argument("max-frames must be positive or -1.");
        }

        cv::VideoCapture capture(inputPath);
        if (!capture.isOpened()) {
            throw std::runtime_error("Could not open input video: " + inputPath);
        }

        cv::Mat frame;
        if (!capture.read(frame) || frame.empty()) {
            throw std::runtime_error("Could not read the first input frame.");
        }

        cv::Rect box = selectRoi
            ? cv::selectROI("Select object", frame, false, false)
            : parseBoundingBox(parser.get<std::string>("bbox"));
        if (selectRoi) {
            cv::destroyWindow("Select object");
        }
        validateBoundingBox(box, frame.size());

        // Every tracker receives the same validated first-frame rectangle.
        cv::Ptr<cv::Tracker> tracker = createTracker(trackerName, modelsDir);
        tracker->init(frame, box);

        double sourceFps = capture.get(cv::CAP_PROP_FPS);
        if (!(sourceFps > 0.0)) {
            sourceFps = 30.0;
        }
        cv::VideoWriter writer;
        if (!outputPath.empty()) {
            writer = openWriter(outputPath, sourceFps, frame.size());
        }

        int framesProcessed = 0;
        int successfulUpdates = 0;
        double elapsedTotal = 0.0;
        cv::Mat lastAnnotated;
        while ((maxFrames < 0 || framesProcessed < maxFrames) &&
               capture.read(frame) && !frame.empty()) {
            const int64 start = cv::getTickCount();
            const bool found = tracker->update(frame, box);
            const double elapsed =
                (cv::getTickCount() - start) / cv::getTickFrequency();
            elapsedTotal += elapsed;
            ++framesProcessed;

            if (found) {
                cv::rectangle(frame, box, cv::Scalar(255, 0, 0), 2);
                ++successfulUpdates;
            } else {
                cv::putText(
                    frame, "Tracking failure detected", cv::Point(20, 40),
                    cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 0, 255), 2);
            }

            const double instantaneousFps =
                elapsed > 0.0 ? 1.0 / elapsed : 0.0;
            std::ostringstream label;
            label << trackerName << " | " << std::fixed
                  << std::setprecision(1) << instantaneousFps << " FPS";
            cv::putText(
                frame, label.str(), cv::Point(20, frame.rows - 20),
                cv::FONT_HERSHEY_SIMPLEX, 0.6, cv::Scalar(50, 170, 50), 2);
            lastAnnotated = frame.clone();

            if (writer.isOpened()) {
                writer.write(frame);
            }
            if (display) {
                cv::imshow("Object tracking", frame);
                const int key = cv::waitKey(1) & 0xff;
                if (key == 27 || key == 'q') {
                    break;
                }
            }
        }

        if (framesProcessed == 0) {
            throw std::runtime_error(
                "The input contains no frames after initialization.");
        }
        if (!snapshotPath.empty() &&
            !cv::imwrite(snapshotPath, lastAnnotated)) {
            throw std::runtime_error(
                "Could not write snapshot: " + snapshotPath);
        }

        std::cout << std::fixed << std::setprecision(4)
                  << "{\"tracker\":\"" << trackerName
                  << "\",\"frames_processed\":" << framesProcessed
                  << ",\"successful_updates\":" << successfulUpdates
                  << ",\"success_rate\":"
                  << static_cast<double>(successfulUpdates) / framesProcessed
                  << ",\"last_bbox\":["
                  << box.x << ',' << box.y << ',' << box.width << ',' << box.height
                  << "],\"frame_size\":[" << frame.cols << ',' << frame.rows
                  << "],\"source_fps\":" << sourceFps
                  << ",\"mean_tracking_ms\":"
                  << 1000.0 * elapsedTotal / framesProcessed << "}\n";
        return 0;
    } catch (const cv::Exception& error) {
        std::cerr << "OpenCV error: " << error.what() << '\n';
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
    }
    return 2;
}

File: tracking/tracker.py
#!/usr/bin/env python3
"""Track one object in a video with OpenCV's modern single-object tracker API."""

from __future__ import annotations

import argparse
import json
import math
from pathlib import Path
from time import perf_counter
from typing import Any

import cv2


# Resolve the default model folder beside this script, independent of cwd.
DEFAULT_MODELS_DIR = Path(__file__).resolve().parent / "models"
# Classic trackers construct without external neural-network weights.
CLASSIC_TRACKER_NAMES = ("MIL", "CSRT", "KCF")
# Each model-backed tracker must receive exactly the files its Params object uses.
MODEL_TRACKER_FILES = {
    "DASIAMRPN": (
        "dasiamrpn_model.onnx",
        "dasiamrpn_kernel_cls1.onnx",
        "dasiamrpn_kernel_r1.onnx",
    ),
    "NANO": (
        "nanotrack_backbone_sim.onnx",
        "nanotrack_head_sim.onnx",
    ),
    "VIT": ("object_tracking_vittrack_2023sep.onnx",),
}
# Class names are kept as data because Python bindings expose two layouts.
MODEL_TRACKER_CLASSES = {
    "DASIAMRPN": "TrackerDaSiamRPN",
    "NANO": "TrackerNano",
    "VIT": "TrackerVit",
}
# Accept the longer names readers commonly use while keeping a compact CLI.
TRACKER_ALIASES = {"NANOTRACK": "NANO", "VITTRACK": "VIT"}
TRACKER_NAMES = (*CLASSIC_TRACKER_NAMES, *MODEL_TRACKER_FILES)


def _normalized_tracker_name(name: str) -> str:
    """Return the canonical command-line name used in summaries and errors."""
    normalized = name.upper()
    return TRACKER_ALIASES.get(normalized, normalized)


def _tracker_creator(class_name: str) -> Any | None:
    """Resolve either spelling used by OpenCV's Python tracker bindings."""
    flat_creator = getattr(cv2, f"{class_name}_create", None)
    if flat_creator is not None:
        return flat_creator
    tracker_class = getattr(cv2, class_name, None)
    if tracker_class is not None:
        return getattr(tracker_class, "create", None)
    return None


def _tracker_params(class_name: str) -> Any | None:
    """Create the Params object exposed by either binding layout."""
    params_class = getattr(cv2, f"{class_name}_Params", None)
    if params_class is None:
        tracker_class = getattr(cv2, class_name, None)
        params_class = (
            getattr(tracker_class, "Params", None)
            if tracker_class is not None
            else None
        )
    return params_class() if params_class is not None else None


def _create_model_tracker(name: str, models_dir: Path) -> Any:
    """Validate model assets, configure the DNN backend, and create a tracker."""
    required_files = MODEL_TRACKER_FILES[name]
    # Check the complete set before OpenCV emits a less actionable load error.
    missing = [
        models_dir / filename
        for filename in required_files
        if not (models_dir / filename).is_file()
    ]
    if missing:
        missing_names = ", ".join(path.name for path in missing)
        raise FileNotFoundError(
            f"{name} requires model files in {models_dir}; missing: "
            f"{missing_names}. Run: python download_models.py "
            f"--tracker {name.lower()}"
        )

    # Resolve both factory spellings used across wheel and source builds.
    class_name = MODEL_TRACKER_CLASSES[name]
    creator = _tracker_creator(class_name)
    params = _tracker_params(class_name)
    if creator is None or params is None:
        raise RuntimeError(
            f"{name} is unavailable in this OpenCV build. Use OpenCV 4.9 or "
            "newer with the video and DNN modules."
        )
    # Tracker classes can be present in a build whose DNN module was disabled.
    dnn = getattr(cv2, "dnn", None)
    if dnn is None:
        raise RuntimeError(
            f"{name} requires an OpenCV build that includes the DNN module."
        )

    # Populate the differently named path fields in each OpenCV Params class.
    if name == "DASIAMRPN":
        params.model = str(models_dir / required_files[0])
        params.kernel_cls1 = str(models_dir / required_files[1])
        params.kernel_r1 = str(models_dir / required_files[2])
    elif name == "NANO":
        params.backbone = str(models_dir / required_files[0])
        params.neckhead = str(models_dir / required_files[1])
    else:
        params.net = str(models_dir / required_files[0])

    # Explicit CPU settings provide the same documented path in OpenCV 4 and 5.
    params.backend = dnn.DNN_BACKEND_OPENCV
    params.target = dnn.DNN_TARGET_CPU
    try:
        return creator(params)
    except cv2.error as exc:
        raise RuntimeError(
            f"OpenCV could not create {name}. Confirm that this build includes "
            "the DNN module and that download_models.py verified every model."
        ) from exc


def create_tracker(name: str, models_dir: Path = DEFAULT_MODELS_DIR) -> Any:
    """Create a classic or model-backed tracker exposed by this OpenCV build."""
    normalized = _normalized_tracker_name(name)
    if normalized not in TRACKER_NAMES:
        raise ValueError(
            f"Unsupported tracker '{name}'. Choose one of: {', '.join(TRACKER_NAMES)}"
        )

    # Neural trackers need explicit, checksum-verified model paths.
    if normalized in MODEL_TRACKER_FILES:
        return _create_model_tracker(normalized, Path(models_dir))

    # Classic factories may be in the main namespace or OpenCV 4's legacy one.
    factory_name = f"Tracker{normalized}_create"
    for namespace in (cv2, getattr(cv2, "legacy", None)):
        if namespace is None:
            continue
        if hasattr(namespace, factory_name):
            return getattr(namespace, factory_name)()
        tracker_class = getattr(namespace, f"Tracker{normalized}", None)
        if tracker_class is not None and hasattr(tracker_class, "create"):
            return tracker_class.create()

    if normalized in {"CSRT", "KCF"}:
        raise RuntimeError(
            f"{normalized} is unavailable in this build. It is a contrib tracker "
            "when exposed and was absent from the exact OpenCV 5.0 builds tested "
            "here; use MIL for the verified cross-version example."
        )
    raise RuntimeError(
        "MIL is unavailable. Install an OpenCV build that includes the video "
        "module's tracker API."
    )


def parse_bbox(value: str) -> tuple[int, int, int, int]:
    """Parse x,y,width,height and reject non-positive boxes."""
    try:
        box = tuple(int(part.strip()) for part in value.split(","))
    except ValueError as exc:
        raise argparse.ArgumentTypeError(
            "The bounding box must contain four integers: x,y,width,height"
        ) from exc
    if len(box) != 4 or box[2] <= 0 or box[3] <= 0:
        raise argparse.ArgumentTypeError(
            "The bounding box must be x,y,width,height with positive dimensions"
        )
    return box


def _checked_bbox(
    bbox: tuple[int, int, int, int], frame_width: int, frame_height: int
) -> tuple[int, int, int, int]:
    x, y, width, height = bbox
    if (
        x < 0
        or y < 0
        or width <= 0
        or height <= 0
        or x + width > frame_width
        or y + height > frame_height
    ):
        raise ValueError(
            f"Bounding box {bbox} lies outside the {frame_width}x{frame_height} frame"
        )
    return bbox


def _open_writer(
    path: Path, fps: float, frame_size: tuple[int, int]
) -> cv2.VideoWriter:
    path.parent.mkdir(parents=True, exist_ok=True)
    codec = "mp4v" if path.suffix.lower() == ".mp4" else "MJPG"
    writer = cv2.VideoWriter(
        str(path),
        cv2.VideoWriter_fourcc(*codec),
        fps,
        frame_size,
    )
    if not writer.isOpened():
        raise RuntimeError(f"Could not open output video: {path}")
    return writer


def run_tracking(
    input_path: Path,
    bbox: tuple[int, int, int, int] | None,
    *,
    tracker_name: str = "MIL",
    models_dir: Path = DEFAULT_MODELS_DIR,
    output_path: Path | None = None,
    snapshot_path: Path | None = None,
    max_frames: int | None = None,
    display: bool = False,
    select_roi: bool = False,
) -> dict[str, Any]:
    """Run tracking and return machine-readable measurements."""
    if max_frames is not None and max_frames <= 0:
        raise ValueError("max_frames must be positive")

    capture = cv2.VideoCapture(str(input_path))
    if not capture.isOpened():
        raise FileNotFoundError(f"Could not open input video: {input_path}")

    writer: cv2.VideoWriter | None = None
    last_annotated = None
    try:
        ok, frame = capture.read()
        if not ok or frame is None:
            raise RuntimeError(f"Could not read the first frame: {input_path}")

        frame_height, frame_width = frame.shape[:2]
        if select_roi:
            selected = cv2.selectROI("Select object", frame, False, False)
            cv2.destroyWindow("Select object")
            bbox = tuple(int(value) for value in selected)
        if bbox is None:
            raise ValueError("Provide --bbox or use --select-roi")
        bbox = _checked_bbox(bbox, frame_width, frame_height)

        tracker = create_tracker(tracker_name, models_dir)
        init_result = tracker.init(frame, bbox)
        if init_result is False:
            raise RuntimeError("Tracker initialization failed")

        source_fps = capture.get(cv2.CAP_PROP_FPS)
        if not math.isfinite(source_fps) or source_fps <= 0:
            source_fps = 30.0
        if output_path is not None:
            writer = _open_writer(
                output_path, source_fps, (frame_width, frame_height)
            )

        frames_processed = 0
        successful_updates = 0
        elapsed_total = 0.0
        last_bbox = bbox

        while max_frames is None or frames_processed < max_frames:
            ok, frame = capture.read()
            if not ok or frame is None:
                break

            start = perf_counter()
            found, updated_bbox = tracker.update(frame)
            elapsed = perf_counter() - start
            elapsed_total += elapsed
            frames_processed += 1

            if found:
                last_bbox = tuple(int(round(value)) for value in updated_bbox)
                x, y, width, height = last_bbox
                cv2.rectangle(
                    frame,
                    (x, y),
                    (x + width, y + height),
                    (255, 0, 0),
                    2,
                )
                successful_updates += 1
            else:
                cv2.putText(
                    frame,
                    "Tracking failure detected",
                    (20, 40),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.7,
                    (0, 0, 255),
                    2,
                )

            instantaneous_fps = 1.0 / elapsed if elapsed > 0 else 0.0
            cv2.putText(
                frame,
                f"{tracker_name.upper()} | {instantaneous_fps:.1f} FPS",
                (20, frame_height - 20),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.6,
                (50, 170, 50),
                2,
            )
            last_annotated = frame

            if writer is not None:
                writer.write(frame)
            if display:
                cv2.imshow("Object tracking", frame)
                if cv2.waitKey(1) & 0xFF in (27, ord("q")):
                    break

        if frames_processed == 0:
            raise RuntimeError("The input contains no frames after initialization")
        if snapshot_path is not None and last_annotated is not None:
            snapshot_path.parent.mkdir(parents=True, exist_ok=True)
            if not cv2.imwrite(str(snapshot_path), last_annotated):
                raise RuntimeError(f"Could not write snapshot: {snapshot_path}")

        return {
            "tracker": _normalized_tracker_name(tracker_name),
            "frames_processed": frames_processed,
            "successful_updates": successful_updates,
            "success_rate": successful_updates / frames_processed,
            "last_bbox": list(last_bbox),
            "frame_size": [frame_width, frame_height],
            "source_fps": source_fps,
            "mean_tracking_ms": 1000.0 * elapsed_total / frames_processed,
        }
    finally:
        capture.release()
        if writer is not None:
            writer.release()
        if display or select_roi:
            cv2.destroyAllWindows()


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--input",
        type=Path,
        default=Path(__file__).parent / "videos" / "chaplin.mp4",
    )
    parser.add_argument(
        "--tracker",
        choices=TRACKER_NAMES,
        default="MIL",
        help=(
            "MIL/CSRT/KCF are model-free; DASIAMRPN/NANO/VIT use ONNX files"
        ),
    )
    parser.add_argument(
        "--models-dir",
        type=Path,
        default=DEFAULT_MODELS_DIR,
        help=f"verified ONNX directory (default: {DEFAULT_MODELS_DIR})",
    )
    parser.add_argument(
        "--bbox",
        type=parse_bbox,
        default=(287, 23, 86, 320),
        help="Initial x,y,width,height (default: 287,23,86,320)",
    )
    parser.add_argument("--select-roi", action="store_true")
    parser.add_argument("--output", type=Path)
    parser.add_argument("--snapshot", type=Path)
    parser.add_argument("--max-frames", type=int)
    parser.add_argument("--display", action="store_true")
    return parser


def main() -> int:
    args = build_parser().parse_args()
    try:
        summary = run_tracking(
            args.input,
            args.bbox,
            tracker_name=args.tracker,
            models_dir=args.models_dir,
            output_path=args.output,
            snapshot_path=args.snapshot,
            max_frames=args.max_frames,
            display=args.display,
            select_roi=args.select_roi,
        )
    except (FileNotFoundError, RuntimeError, ValueError) as exc:
        print(f"error: {exc}")
        return 2
    print(json.dumps(summary, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
